### ISSUE TITLE
`-C check-cfg` emits compile errors

### ISSUE THREAD

**Issue by daxpedda** [katharostech/cfg_aliases#7]
With `-C check-cfg` stabilized in Cargo v1.80 and enabled by default (see https://github.com/rust-lang/cargo/pull/13571), newly introduced `cfg`s will emit compile errors when used inside the crate.

The solution is to emit `cargo:rustc-check-cfg=cfg(your_cfg)` inside the `build.rs` to register the new `cfg`s, this will emit warnings in older versions of Rust because the feature is not stable, but at least it won't emit errors in Rust v1.80 (or currently nightly).

See the [`-C check-cfg`](https://doc.rust-lang.org/nightly/rustc/check-cfg.html) and the accompanying [`rustc-check-cfg`](https://doc.rust-lang.org/nightly/cargo/reference/build-scripts.html?highlight=rustc-check-cfg#rustc-check-cfg) documentation.

**Comment by zicklag** [comment]
I haven't checked the best way to do this yet, but do you think it's worth conditioning on the rustc version to avoid causing warnings in earlier versions of Rust? I wouldn't want this to mess up people who have CI set to deny warnings or other similar situations.

**Comment by zicklag** [comment]
It looks like this is basically the best way to get the Rust version for now, taken from the semver crate [here](https://github.com/dtolnay/semver/blob/master/build.rs):

```rust
fn rustc_minor_version() -> Option<u32> {
    let rustc = env::var_os("RUSTC")?;
    let output = Command::new(rustc).arg("--version").output().ok()?;
    let version = str::from_utf8(&output.stdout).ok()?;
    let mut pieces = version.split('.');
    if pieces.next() != Some("rustc 1") {
        return None;
    }
    pieces.next()?.parse().ok()
}
```

I wish it didn't involve calling out to `rustc`, but it's not too bad.

@ogoffart thanks for the PR! Do you think you could add a Rustc version check so we can avoid showing unnecessary warnings?

**Comment by ogoffart** [comment]
IMHO this is overkill to add the check.
From my understanding, the warning occurs for versions ≤ 1.76.
1.77 is the previous stable already. Not that old, but according to the [2023 survey](https://raw.githubusercontent.com/rust-lang/surveys/main/surveys/2023-annual-survey/report/annual-survey-2023-report.pdf) this account for more than 90% of developers.
The resulting warning with old version is a cargo warning which cannot be turned into an error and can never make CI fails.
And other part of the ecosystem such as winit, nix, and so on, do it unconditionally anyway.

**Comment by Urgau** [comment]
> Not that old, but according to the [2023 survey](https://raw.githubusercontent.com/rust-lang/surveys/main/surveys/2023-annual-survey/report/annual-survey-2023-report.pdf) this account for more than 90% of developers.

Note that if you want more accurate data crates.io had a public [dashboard of the download requests by cargo version](https://p.datadoghq.com/sb/3a172e20-e9e1-11ed-80e3-da7ad0900002-973f4c1011257befa8598303217bfe3a?fromUser=false&refresh_mode=sliding&from_ts=1714638517653&to_ts=1715243317653&live=true). It in versions >=1.77 currently represent ~67% of all the download (CI, users, automation, ...).

I therefore agree with your analysis that there is no need to condition the warnings in relation to a version of Cargo/rustc.

**Comment by zicklag** [comment]
> which cannot be turned into an error and can never make CI fails

Ah, OK, in that case I agree there's no reason to do the check.

**Comment by zicklag** [comment]
Merged! I'm going to test this locally later today, and then publish a new version of the crate. Thank you! 👍️

**Comment by zicklag** [comment]
The fix has been published in version `0.2.1` ✅
